core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -421,7 +423,32 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     // the callback for sending a fetch response
     def sendResponseCallback(responsePartitionData: Map[TopicAndPartition, FetchResponsePartitionData]) {
-      val mergedResponseStatus = responsePartitionData ++ unauthorizedResponseStatus
+
+      val convertedResponseStatus =
+        // Need to down-convert message when consumer only takes magic value 0.
+        if (fetchRequest.versionId <= 1) {
+          responsePartitionData.map({ case (tp, data) =>
+            tp -> {
+              // We only do down-conversion when:
+              // 1. The message format version configured for the topic is using magic value > 0, and
+              // 2. The message set contains message whose magic > 0
+              // This is to reduce the message format conversion as much as possible. The conversion will only occur
+              // when new message format is used for the topic and we see an old request.
+              // Please notice that if the message format is changed from a higher version back to lower version this
+              // test might break because some messages in new message format can be delivered to consumer without
+              // format down conversion.
+              if (replicaManager.getMessageFormatVersion(tp).exists(_ > 0) &&
```

Comment:
Thanks, got your point. So, setting the message version to 0 in the config not only implies that future messages will be written in version 0, but all existing messages are of version 0 too? We should at least document this, but I am not sure if this is enough to warn people about the impact of switching back and forth of the message format config.

Could you run some experiments about the performance impact of doing the message format check?